app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -162,19 +180,54 @@ func (s *Hijacker) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
 		close(out)
 	}()
 
+	// Tracks unique content type priorities returned by applications in sorted order.
+	// As there are only a few content types supported by prometheus scraper,
+	// we can use a simple array instead of a map.
+	ctPriorities := make([]bool, len(prometheusContentTypePriorities))
 	for _, app := range s.applicationsToScrape {
 		go func(app ApplicationToScrape) {
 			defer wg.Done()
-			out <- s.getStats(ctx, req, app)
+			content, contentType := s.getStats(ctx, req, app)
+			out <- content
+			if content != nil && contentType != "" {
+				// skip unknown content type
+				priorityIdx, valid := prometheusContentTypePriorities[contentType]
+				if !valid {
+					return
+				}
+				// if contentType is not found in the priorities map, it will return 0
+				ctPriorities[priorityIdx] = true
+			}
 		}(app)
 	}
 
 	select {
 	case <-ctx.Done():
 		return
 	case <-done:
-		// default format returned by prometheus
-		writer.Header().Set("content-type", string(expfmt.FmtText))
+		// We should not simply use the highest priority content type even if `application/openmetrics-text`
+		// is the superset of `text/plain`, as it might not be
+		// supported by the applications or the user might be using older prom scraper.
+		// So it's better to choose the highest negotiated content type between the apps and the scraper.
+		ct := ""
+		for priority := range ctPriorities {
+			if ctPriorities[priority] {
+				ct = prometheusPriorityContentType[priority]
+				break
+			}
+		}
```

Comment:
if we always `break` on the first one we find, I think we don't need a `[]bool`.

Reply:
You only need to keep track of the highest priority content type seen, you don't need to keep a list of every type seen and then pick the first in the list.

Reply:
Oh my Gosh! that was so silly.. I will fix that.